### PR TITLE
#29 Block sparse GEMM implementation

### LINKED ISSUE #29: Implement Block SpGEMM
# TODO:

* [ ] Tests:
  - [x] simple explicit matrices with known multiplication result;
  - [x] implement block version of random matrix generator;
  - [ ] address regression test for [#402](https://github.com/kokkos/kokkos-kernels/issues/402): implement or drop.
* [ ] Interface: can we conveniently avoid separate `bspgemm` methods with proper templating over `spgemm` ?
* [ ] Address review feedback and tune performance;

## Implementation

* [x] Serial/Debug algorightm implementation (`SPGEMM_DEBUG`);
* [ ] Default algorithms Serial/Debug algorightm implementation (`SPGEMM_KK_MEMORY` and `SPGEMM_KK_MEMSPEED`);
  - [x] adjust hash accumulator for blocks;
  - [x] cover CPU variants: `MultiCoreTag` and `MultiCoreTag4`  (`SPGEMM_KK_LP`);
  - [ ] adjust memory estimation and algorithm/execution policy selection for block version;
    - [x] CPU (reduced to `suggested_vector_size`, `suggested_team_size`, and `team_row_chunk_size`);
    - [ ] GPU - detailed review of all estimations and allocations (+refactoring if possible).
  - [ ] GPU implementation;
  - [ ] Can we get unified accumulator (like `HashmapAccumulator` for non-block version);
* [ ] Connect TPLs for **performance** reference:
  - [x] _MKL_ for CPU benchmarking;
  - [ ] _CuSPARSE_ for GPU benchmarking;
* [ ] Speed numeric method: `SPGEMM_KK_SPEED`;

### PR DESCRIPTION
**Pull request by mzuzek**
Fixes #29

Mirror PR in the upstream repository: https://github.com/kokkos/kokkos-kernels/pull/1099

### CODE REVIEW

**Comment by mzuzek** [comment]
@uhetmaniuk @fnrizzi 

I got CPU version of the default algorithm working:
* Symbolic version is reused from non-block `spgemm` without modification.
* Numeric phase was implemented `KokkosBSPGEMM` in _sparse/impl/KokkosSparse_bspgemm_impl_kkmem.hpp_ with `MultiCoreTag` (`SPGEMM_KK`) and `MultiCoreTag4`  (`SPGEMM_KK_LP`) functors.
* Accumulator is work-in-progress: `MultiCoreTag` operator uses `HashingIndex` which is essentially `HashmapAccumulator` reduced to indexing used by spgemm (`HashmapAccumulator` works on scalars) while `MultiCoreTag4` operator inlines the indexing like it did in non-block version. I'll try to refactor/unify it once I dig into GPU operators and understand all indexing/accumulation scenarios. This plan includes detailed review of related memory estimations and allocations.
* For block accumulation/combination I used `KokkosBatched::SerialGemmInternal<KokkosBatched::Algo::Gemm::Unblocked>`  from _sparse/impl/KokkosBatched_Gemm_Serial_Internal.hpp_ (`C = A * B` and `C += A * B` with `beta` of 0 and 1 respectively).
* Unit tests are based on new overload of `kk_generate_sparse_matrix()`  in _common/KokkosKernels_IOUtils.hpp_ that takes block size and generates random BSR matrix (random CSR and BSR content is generated by shared `generate_bsr_content()` routine, I also fixed couple of corner cases there, so it can generate small matrices as well).
* There are some extra utilities for unit testing in _unit_test/sparse/Test_Sparse_bspgemm.hpp_, including `<<` operator that renders BSR matrix on `stdout`.

Edit Sep 9, 2021: added note on batched dgemm for block accumulation

**Comment by mzuzek** [comment]
@uhetmaniuk @fnrizzi 

I connected MKL to benchmark block spgemm performance on CPU.

> Note: In non-block version, `mkl_sparse_spmm()`  is called twice: in symbolic and numeric phases. I drafted a new interface that avoids this duplicated allocations and calculation to get more accurate insight into performance baseline. This optimization would need some discussion to be included in the actual implementation (i.e. outside the unit test), since it doesn't immediately fit the interface, which forces separate symbolic and numeric phases.

For multiplication of random matrix by itself on [i9-9900K](https://www.intel.com/content/www/us/en/products/sku/186605/intel-core-i99900k-processor-16m-cache-up-to-5-00-ghz/specifications.html) (8 cores/16 threads)/32GB RAM/Ubuntu/gcc-8.4 (WSL), I got some rough timings (no repetition, accuracy around 0,5~1s):

```
Note: Google Test filter = *block_spgemm*
[==========] Running 1 test from 1 test case.
[----------] Global test environment set-up.
[----------] 1 test from threads
[ RUN      ] threads.sparse_block_spgemm_double_int_int_TestExecSpace

Testing BSPGEMM: nrows=300000   block:2x2       nnz=8000000(0.009%)     bandwidth=100000(33.3%)        variance=10 (0.0%)      schedule=dynamic
 nnz_in=7800201
nnz_out=202326755
SPGEMM_DEBUG (reference result): 17.0
SPGEMM_MKL              spgemm_time:    5.5     output_check_time:      0.6
SPGEMM_KK               spgemm_time:    1.7     output_check_time:      0.5
SPGEMM_KK_LP            spgemm_time:    1.9     output_check_time:      0.5

Testing BSPGEMM: nrows=300000   block:3x3       nnz=6000000(0.007%)     bandwidth=100000(33.3%)        variance=10 (0.0%)      schedule=dynamic
 nnz_in=6000201
nnz_out=119850101
SPGEMM_DEBUG (reference result): 23.0
SPGEMM_MKL              spgemm_time:    10.7    output_check_time:      6.7
SPGEMM_KK               spgemm_time:    4.8     output_check_time:      0.7
SPGEMM_KK_LP            spgemm_time:    2.2     output_check_time:      0.7
[       OK ] threads.sparse_block_spgemm_double_int_int_TestExecSpace (80583 ms)
[----------] 1 test from threads (80587 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test case ran. (80589 ms total)
[  PASSED  ] 1 test.
```
> Note: I tried to measure largest matrices I could fit into memory. For reference: _audikw_1_ - our largest [_spmv_ sample](https://github.com/fnrizzi/kokkos-kernels/issues/3#issuecomment-842759224) has around 40e+6 non-zero values.

> Note: _nnz_ means non-zero **blocks**.

> Note: Tested on _Threads_ Kokkos execution space (because of silly issue with _OpenMP_ running on single core...)

Would that look reasonable - based on your spgemm perf experience/expectations ?
Or should we run more tests ?

File: src/common/KokkosKernels_IOUtils.hpp [hunk 1]
```diff
@@ -413,61 +412,113 @@ crsMat_t kk_generate_triangular_sparse_matrix(
...
+    bandwidth, rowmap_view, columns_view, values_view);
+
+  typename crsMat_t::StaticCrsGraphType static_graph(columns_view, rowmap_view);
   crsMat_t crsmat("CrsMatrix", ncols, values_view, static_graph);
-  delete [] xadj; delete [] adj; delete [] values;
   return crsmat;
 }
 
+template <typename bsrMat_t>
+bsrMat_t kk_generate_sparse_matrix(
```
Comment (uhetmaniuk) [inline src/common/KokkosKernels_IOUtils.hpp:450]: Naive question: could you fill first a CrsMatrix and then use the BsrMatrix constructor that takes a CrsMatrix and a block dimension as input parameters?

**Comment by uhetmaniuk** [comment]
Looks good. I do not have any other comments.

File: src/common/KokkosKernels_IOUtils.hpp (hunk 1 above)
Comment (mzuzek) [reply src/common/KokkosKernels_IOUtils.hpp:450]: Yes, I reverted to this simpler approach

**Comment by mzuzek** [comment]
@uhetmaniuk @fnrizzi 
Thank you for the review. I opened draft [PR#1099](https://github.com/kokkos/kokkos-kernels/pull/1099) in the KK repository.

**Comment by mzuzek** [comment]
Closing this internal version - please comment on regular PR https://github.com/kokkos/kokkos-kernels/pull/1099

**Pull request closed without merging**